go/WorkQueue.go

```diff
@@ -120,6 +122,67 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
 	return err
 }
 
+func (workQueue *WorkQueue) AddAtomicItem(ctx context.Context, db *redis.Client, item Item) error {
+	tx := db.TxPipeline()
+
+	// Add LRANGE commands to the transaction
+	tx.LRange(ctx, workQueue.mainQueueKey, 0, -1)
+	tx.LRange(ctx, workQueue.processingKey, 0, -1)
+
+	// Execute the transaction
+	results, err := tx.Exec(ctx)
+	if err != nil {
+		return err
+	}
+
+	mainItems := results[0].(*redis.StringSliceCmd).Val()
+	processingItems := results[1].(*redis.StringSliceCmd).Val()
+	itemID := item.ID
+
+	if sliceContainsString(mainItems, itemID) || sliceContainsString(processingItems, itemID) {
+		// Same item tried being added twice.
+		return nil
+	}
+
+	pipeline := db.TxPipeline()
+	workQueue.AddItemToPipeline(ctx, pipeline, item)
+	_, err = pipeline.Exec(ctx)
+	return err
+}
+
+func (workQueue *WorkQueue) GetQueueLengthsAtomic(ctx context.Context, db *redis.Client) ([]int64, error) {
+	tx := db.TxPipeline()
+
+	queueLen := tx.LLen(ctx, workQueue.mainQueueKey)
+	processingLen := tx.LLen(ctx, workQueue.processingKey)
+
+	_, err := tx.Exec(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]int64, 2)
+	result[0], _ = queueLen.Result()
+	result[1], _ = processingLen.Result()
+
+	return result, nil
+}
+
+func (workQueue *WorkQueue) GetQueueLengths(ctx context.Context, db *redis.Client) ([]int64, error) {
+	pipe := db.Pipeline()
+	queueLen := pipe.LLen(ctx, workQueue.mainQueueKey)
+	processingLen := pipe.LLen(ctx, workQueue.processingKey)
+	_, err := pipe.Exec(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]int64, 2)
+	result[0], _ = queueLen.Result()
+	result[1], _ = processingLen.Result()
+	return result, err
+}
```

Comment:
I don't think there's a need for the non-atomic method. If you want the queue lengths non-atomically, then you can call the existing separate methods. If you want them both, I think you'd generally assume it's atomic, so I think `GetQueueLengths` should be the transaction version, and that should be the only version?

Also, on the naming, we currently have:
- `QueueLen`
- `Processing`

It might be a more consistent name to call it `Lengths`? Perhaps just `Len` but that might be unclear? What's your opinion?